src/providers/originProvider.ts

```diff
@@ -0,0 +1,105 @@
+import { CACHE_HEADERS } from '../constants/cache';
+import { Context } from '../context';
+import {
+  GetFileOptions,
+  GetFileResult,
+  HeadFileResult,
+  HttpResponseHeaders,
+  Provider,
+  ReadDirectoryResult,
+} from './provider';
+
+type OriginProviderCtorOptions = {
+  ctx: Context;
+};
+
+/**
+ * Serves assets from origin.nodejs.org, used as a fallback for if R2 fails.
+ */
+export class OriginProvider implements Provider {
+  private ctx: Context;
+
+  constructor({ ctx }: OriginProviderCtorOptions) {
+    this.ctx = ctx;
+  }
+
+  async headFile(path: string): Promise<HeadFileResult | undefined> {
+    const res = await fetch(this.ctx.env.ORIGIN_HOST + path, {
+      method: 'HEAD',
+      headers: {
+        'user-agent': 'release-cloudflare-worker',
+      },
+    });
+
+    if (res.status === 404) {
+      return undefined;
+    }
+
+    return {
+      httpStatusCode: res.status,
+      httpHeaders: originHeadersToOurHeadersObject(res.headers),
+    };
+  }
+
+  async getFile(
+    path: string,
+    options?: GetFileOptions | undefined
+  ): Promise<GetFileResult | undefined> {
+    const res = await fetch(this.ctx.env.ORIGIN_HOST + path, {
+      headers: {
+        'user-agent': 'release-cloudflare-worker',
```

Comment:
Interesting, love the approach on the cache headers by using content matches, so that the browser can "proactively" fetch and replace on successive calls to this endpoint if an update was done 🆒